daniel/pass/ps.c

```diff
@@ -0,0 +1,83 @@
+#include "ps.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <dirent.h>
+
+int_array getProcesses() {
+    DIR *directory;
+    int_array pidList;
+    int counter = 0;
+    struct dirent *file;
+    directory = opendir("/proc");
+    if (directory) {
+        while ((file = readdir(directory)) != NULL) {
+            if (atoi(file->d_name)) {
+                pidList.data[counter++] = atoi(file->d_name);
+            }
+        }
+        pidList.length = counter;
+        closedir(directory);
+    }
+    return pidList;
+}
+
+void replace(char *string, int length, char old, char new) {
+    for (int counter = 1; counter < length; counter++) {
+        if (string[counter] == old) {
+            string[counter] = new;
+        }
+    }
+    string[length - 1] = 0;
+}
+
+void getProcessName(Process *process, int pid) {
+    char statFileName[MAX_PROC_PATH];
+    sprintf(statFileName, "/proc/%d/stat", pid);
+    FILE *statFile = fopen(statFileName, "r");
+    fscanf(statFile, "%*s %s", process->name);
+}
+
+void getProcessCommand(Process *process, int pid) {
+    char cmdFileName[MAX_PROC_PATH];
+    char *cmdData = (char *) calloc(MAX_COMMAND, sizeof(char));
+    int copiedLength = 0;
+    sprintf(cmdFileName, "/proc/%d/cmdline", pid);
+    FILE *fdCmdLine = fopen(cmdFileName, "r");
+
+    if (fdCmdLine) {
+        copiedLength = fread((void *) cmdData, sizeof(char), MAX_COMMAND, fdCmdLine);
+        fclose(fdCmdLine);
+        if (copiedLength) {
+            replace(cmdData, copiedLength, '\0', ' ');
+            strcpy(process->command, cmdData);
+        } else {
+            strcpy(process->command, "null");
+        }
+    } else {
+        strcpy(process->command, "null");
+    }
+    free(cmdData);
+}
+
+void getProcessPath(Process *process, int pid) {
+    char *binaryFile;
+    char realPath[PATH_MAX + 1];
+    sprintf(binaryFile, "/proc/%d/exe", pid);
+    if (readlink(binaryFile, realPath, PATH_MAX) != -1) {
+        strcpy(process->path, realPath);
+    } else {
+        strcpy(process->path, "null");
+    }
+}
+
+Process getProcessInfo(int pid) {
+    Process process;
+    getProcessName(&process, pid);
+    getProcessPath(&process, pid);
+    getProcessCommand(&process, pid);
+    process.pid = pid;
+
+    return process;
```

Comment:
I think the struct is allocated on the stack and once you exit the function, the data could be jibberish (but maybe my c is a bit rusty).
If I'm right this could cause problems later on.

Reply:
i checked before and you can and it does work - it can return a struct through the stack.

Reply:
It is on the stack and its quite bad practice to do so.
In this occasion it works, but on a different one  where you call this function inside another one, and send the pointer throught the parent function it will be bad.

It's better to change the funcrion so it recieves a pointer and you allocate it on the heap.

Also, when you write functions that returns a struct, you have little freedom with returning an error value, and it is less readable.

Reply:
i changed it, but can you elaborate on the occasion where my struct's memory can be overwritten?
becuase I am not sure I fully understand it.

Reply:
Well I looked again and I was mistakes that the Process struct is a pointer, but it isn't.
So in this case it doesn't "really" matter, but it is quite better to allocate at the heap (because of the reasons I mentioned before).

Its' hard to describe it here - I'll do it in person when we meet :)